server/src/main/java/org/elasticsearch/ExceptionsHelper.java

```diff
@@ -97,6 +97,195 @@ public static String stackTrace(Throwable e) {
         return stackTraceStringWriter.toString();
     }
 
+    /**
+     * Constructs a limited and compressed stack trace string. Each exception printed as part of the full stack trace will have its printed
+     * stack frames capped at the given trace depth. Stack traces that are longer than the given trace depth will summarize the count of the
+     * remaining frames at the end of the trace. Each stack frame omits the module information and limits the package names to single
+     * characters per part.
+     * <br><br>
+     * An example result when using a trace depth of 2 and one nested cause:
+     * <pre><code>
+     * o.e.s.GenericException: some generic exception!
+     *   at o.e.s.SomeClass.method(SomeClass.java:100)
+     *   at o.e.s.SomeOtherClass.earlierMethod(SomeOtherClass.java:24)
+     *   ... 5 more
+     * Caused by: o.e.s.GenericException: some other generic exception!
+     *   at o.e.s.SomeClass.method(SomeClass.java:115)
+     *   at o.e.s.SomeOtherClass.earlierMethod(SomeOtherClass.java:16)
+     *   ... 12 more
+     * </code></pre>
+     *
+     * @param e Throwable object to construct a printed stack trace for
+     * @param traceDepth The maximum number of stack trace elements to display per exception referenced
+     * @return A string containing a limited and compressed stack trace.
+     */
+    public static String limitedStackTrace(Throwable e, int traceDepth) {
+        assert traceDepth >= 0 : "Cannot print stacktraces with negative trace depths";
+        StringWriter stackTraceStringWriter = new StringWriter();
+        PrintWriter printWriter = new PrintWriter(stackTraceStringWriter);
+        printLimitedStackTrace(e, printWriter, traceDepth);
+        return stackTraceStringWriter.toString();
+    }
+
+    /** Caption for labeling causative exception stack traces */
+    private static final String CAUSE_CAPTION = "Caused by: ";
+    /** Caption for labeling suppressed exception stack traces */
+    private static final String SUPPRESSED_CAPTION = "Suppressed: ";
+
+    private static void printLimitedStackTrace(Throwable e, PrintWriter s, int maxLines) {
+        // Guard against malicious overrides of Throwable.equals by
+        // using a Set with identity equality semantics.
+        Set<Throwable> dejaVu = Collections.newSetFromMap(new IdentityHashMap<>());
+        dejaVu.add(e);
+        synchronized (s) {
```

Comment:
This is not clear to me either, but I can see it's a conscious choice, so it would be nice to add a line or two to explain the purpose of this.

Reply:
This was left over from the original exception method. Though it looks like the reasoning for this is because you could have passed in a writer that was shared across threads. I'll remove this since it should be safe and self contained.